File: python/Greedy/jump_game.py
from typing import List

class Solution:
    def canJump(self, nums: List[int]) -> bool:
        """
        Determine if you can reach the last index starting from index 0.
        
        Approach:
        - Start from the last index and work backwards.
        - For each position, check if you can reach the current goal from that position.
        - If you can reach the goal from position i, update the goal to i.
        - If the goal becomes 0, you can reach the last index.
        
        Time Complexity: O(N), where N is the length of nums
        - We iterate through the array once from the end
        
        Space Complexity: O(1)
        - Only a constant amount of extra space is used
        
        Args:
            nums: List[int] - List of non-negative integers representing maximum jump length
        Returns:
            bool - True if you can reach the last index, False otherwise
        """
        goal = len(nums) - 1

        for i in range(len(nums) - 1, -1, -1):
            # If you can reach the current goal from position i, update the goal to i
            if i + nums[i] >= goal:
                goal = i
        
        # If the goal is 0, you can reach the last index
        return True if goal == 0 else False

# -------------------
# Test Cases
# -------------------
def test_can_jump():
    s = Solution()
    # Test case 1: Can reach the end
    assert s.canJump([2,3,1,1,4]) == True, "Failed on [2,3,1,1,4]"
    # Test case 2: Cannot reach the end
    assert s.canJump([3,2,1,0,4]) == False, "Failed on [3,2,1,0,4]"
    # Test case 3: Single element
    assert s.canJump([0]) == True, "Failed on [0]"
    # Test case 4: All zeros except last
    assert s.canJump([0,0,0,1]) == False, "Failed on [0,0,0,1]"
    # Test case 5: Can jump directly to end
    assert s.canJump([5,0,0,0]) == True, "Failed on [5,0,0,0]"
    # Test case 6: Large jumps
    assert s.canJump([2,0,0]) == True, "Failed on [2,0,0]"
    # Test case 7: Cannot jump from first position
    assert s.canJump([0,1,2]) == False, "Failed on [0,1,2]"
    # Test case 8: Multiple paths
    assert s.canJump([1,2,3,4,5]) == True, "Failed on [1,2,3,4,5]"
    print("All test cases passed!")

if __name__ == "__main__":
    test_can_jump()

File: C++/Greedy/jump_game.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    /**
     * Function to determine if you can reach the last index starting from index 0.
     *
     * Approach:
     * - Start from the end of the array.
     * - At each index, check if you can reach the current `goal` from there.
     * - If yes, update `goal` to current index.
     * - If at the end `goal == 0`, then you can reach the last index.
     *
     * Time Complexity: O(N) — traverse the array once.
     * Space Complexity: O(1) — uses only constant extra space.
     */
    bool canJump(vector<int>& nums) {
        int goal = nums.size() - 1;  // Initialize goal as the last index

        for (int i = nums.size() - 1; i >= 0; --i) {
            if (i + nums[i] >= goal) {
                // If you can reach `goal` from index `i`, update goal
                goal = i;
            }
        }

        // If goal has been moved to index 0, return true
        return goal == 0;
    }
};

/*
====================================================
Line-by-line C++ syntax explanation:
====================================================

#include <bits/stdc++.h>
    -> Includes all standard C++ libraries (like <vector>, <string>, <iostream>, etc.)

using namespace std;
    -> Allows using standard library names (like vector, string) without prefixing them with std::

class Solution {
public:
    -> Defines a class named Solution with public member functions.

bool canJump(vector<int>& nums) {
    -> Defines a method `canJump` that takes a reference to a vector of integers and returns a bool.

int goal = nums.size() - 1;
    -> Initializes `goal` to the last index of the array.

for (int i = nums.size() - 1; i >= 0; --i) {
    -> Loop from the last index down to index 0.

if (i + nums[i] >= goal) {
    -> If you can reach `goal` from index `i`, update `goal` to `i`.

goal = i;
    -> Updates `goal` to the current index.

return goal == 0;
    -> If `goal` reaches 0, it means you can reach the end starting from 0.

====================================================
Time Complexity:
    - O(N): Iterates over the array exactly once.

Space Complexity:
    - O(1): Only uses a single integer variable `goal`.

====================================================
*/

